### PR TITLE
serializers: added schema processors

### PR DESCRIPTION
**Pull request by alejandromumo**
closes https://github.com/inveniosoftware/invenio-rdm-records/issues/1231

### CODE REVIEW

File: flask_resources/serializers/base.py
Comment (alejandromumo) [inline flask_resources/serializers/base.py:53]: In order to do this, there must a be a refactor in other schemas in `flask-resources`, e.g. `BaseObjectSchema`:

https://github.com/inveniosoftware/flask-resources/blob/d849ba66b151e42e790caa57c642a0c543f6f9a3/flask_resources/parsers/schema.py#L72-L73

File: flask_resources/serializers/json.py
Comment (alejandromumo) [inline flask_resources/serializers/json.py:47]: This change was neeed due to adding positional arguments from `**kwargs` before. Might not be needed if we change the way the schema is initialised.

File: flask_resources/serializers/simple.py
Comment (max-moser) [inline flask_resources/serializers/simple.py:16]: Just wondering, not a blocker in any way:
Should we maybe store ignored arguments in a field like `self._ignored_kwargs = kwargs`?
Could that help somebody(*) with debugging, in case some keyword arguments are supplied but simply ignored?
Would that cause issues like wasted memory, or maybe even block something (idk, maybe some database models or caches or something)?

(*) most likely that would only ever benefit sysadmins who are comfortable wading through `invenio shell`...

**Review by max-moser — APPROVED** [review]
LGTM!
These changes also make my PR obsolete (which is a good thing for customizability): https://github.com/inveniosoftware/invenio-rdm-records/pull/1266

File: flask_resources/serializers/base.py
Comment (slint) [inline flask_resources/serializers/base.py:56]: minor: now that I look at this, shouldn't we also have below:

```diff
-self.format_serializer_cls = format_serializer_cls(**serializer_options)
+self.format_serializer = format_serializer_cls(**serializer_options)
```

Either that, or we apply the same principle, and store just the classes in `self.format_serializer_cls` and `self.object_schema_cls` and kwargs separately so that they're initialized and used in the methods below... In any case, I would aim for consistency :)

**Review by slint — COMMENTED** [review]
LGTM, one minor comment regarding consistency of storing classes + kwargs, vs. initializing them in `__init__`.

I see some benefits in initializing in the constructor (🅰️):

- Validation: if the parameters are wrong, we'll get a syntax error
- Performance/memory: we only initialize the objects once

Benefits for storing classes/kwargs and initializing later on use (🅱️):

- Better inspection when debugging a serializer
- Sub-classes might be able to customize a bit easier and modify/prepare the classes/kwargs... Then there's also https://github.com/inveniosoftware/invenio-rdm-records/issues/1267 to address at some point...

I feel more in favor of 🅰️ so far, but I'll let other maybe to chime-in.

File: flask_resources/serializers/base.py
Comment (max-moser) [reply flask_resources/serializers/base.py:56]: Good catch, the `self.format_serializer_cls` is actually not a class but an object :eyes: 

Personally, I'd also go for initializing objects in the constructor, rather than cobbling them together every time we call a method like `dumps()`.
Might just be me, but I find myself trying to figure out the call stacks of certain object methods via `invenio shell` rather often, and then the re-construction of objects in every method is just noise to me.
If you wanna fetch the classes for debugging purposes, `type(self.format_serializer)` or `self.format_serializer.__class__` should still work fine.

Comment (alejandromumo) [reply flask_resources/serializers/base.py:56]: Personally I also preffer to initialize it, if possible, in the constructor. Or at least have consistency in general.

File: flask_resources/serializers/simple.py
Comment (alejandromumo) [reply flask_resources/serializers/simple.py:16]: That's a good question. Sometimes we have arguments "leaking" that end up in kwargs 

I am not sure however how much of a performance loss or memory overhead this causes.

File: flask_resources/serializers/base.py
Comment (slint) [reply flask_resources/serializers/base.py:56]: The people have spoken:

- I think we go with initializing in the constructor
- name the class attributes correctly, i.e. without the `_cls` suffix
  - for this, we need to check if it's a breaking change actually for the downstream serializers... if it is then I would open a separate issue and tackle separately

**Comment by ntarocco** [comment]
> 

@max-moser does it mean that you will close it?

File: flask_resources/serializers/base.py
Comment (slint) [reply flask_resources/serializers/base.py:67]: `object_schema_cls ` is indeed used at https://github.com/inveniosoftware/flask-resources/blob/08eab53b3b460c9690203a8465b03c9a415e62c4/flask_resources/parsers/schema.py#L59 for serializing each object in the list schema... The `self.context` passed is the `ctx` built in here. Is it safe to assume that the item and list `context` would always be the same? From a quick look around, it seems to be the case. 

Before merging this I think we have make sure that `invenio-rdm-records` tests are passing with these changes.

Comment (max-moser) [reply flask_resources/serializers/base.py:67]: Looking good for me:
```
[...]
======================================================================================= 642 passed, 782 skipped, 104076 warnings in 1149.82s (0:19:09) ========================================================================================
[+] Running 4/4
 ✔ Container docker_services_cli-redis-1       Removed0.5s  
 ✔ Container docker_services_cli-postgresql-1  Removed0.7s  
 ✔ Container docker_services_cli-opensearch-1  Removed0.9s  
 ✔ Network docker_services_cli_default         Removed0.1s  
(invenio-rdm-records) mmoser@Alpha ~/rdm/modules/invenio-rdm-records [master] $ pip list | grep -i home                                                                                                                             [20:52:03]
flask-resources               1.0.0     /home/mmoser/rdm/modules/flask-resources
invenio-rdm-records           2.11.0    /home/mmoser/rdm/modules/invenio-rdm-records
(invenio-rdm-records) mmoser@Alpha ~/rdm/modules/invenio-rdm-records [master] $ 
```

File: flask_resources/serializers/simple.py
Comment (ppanero) [reply flask_resources/serializers/simple.py]: it does not seem to be necessary with the latest changes. there will also be another big refactor as part of https://github.com/inveniosoftware/flask-resources/issues/117, could be re-evaluated then

File: flask_resources/serializers/base.py
Comment (ppanero) [inline flask_resources/serializers/base.py:116]: This is 100% not ideal, but is the only way to move to this module the processing pipelines. When https://github.com/inveniosoftware/flask-resources/issues/117 is addressed this will be moved to the transformer class.

**Pull request merged**
